python/cudf/cudf/_lib/interop.pyx

```diff
@@ -84,41 +87,63 @@ cdef void dlmanaged_tensor_pycapsule_deleter(object pycap_obj):
     dlpack_tensor.deleter(dlpack_tensor)
 
 
-cdef vector[column_metadata] gather_metadata(object metadata) except *:
+cdef vector[column_metadata] gather_metadata(dict cols_dtypes) except *:
     """
-    Metadata is stored as lists, and expected format is as follows,
-    [["a", [["b"], ["c"], ["d"]]],       [["e"]],        ["f", ["", ""]]].
+    Generates a column_metadata vector for each column.
+
+    Parameters
+    ----------
+    cols_dtypes : dict
+        A dict mapping of column names & their dtypes.
+
     First value signifies name of the main parent column,
     and adjacent list will signify child column.
     """
     cdef vector[column_metadata] cpp_metadata
-    if isinstance(metadata, list):
-        cpp_metadata.reserve(len(metadata))
-        for i, val in enumerate(metadata):
-            cpp_metadata.push_back(column_metadata(str.encode(str(val[0]))))
-            if len(val) == 2:
-                cpp_metadata[i].children_meta = gather_metadata(val[1])
-
-        return cpp_metadata
+    cpp_metadata.reserve(len(cols_dtypes))
+
+    if cols_dtypes is not None:
+        for idx, (col_name, col_dtype) in enumerate(cols_dtypes.items()):
+            cpp_metadata.push_back(column_metadata(col_name.encode()))
+            if is_struct_dtype(col_dtype):
+                _set_col_children_metadata(col_dtype, cpp_metadata[idx])
+    else:
+        raise TypeError(
+            "Requires dtype to be passed to "
+            "construct column_metadata"
+        )
+    return cpp_metadata
+
+cdef _set_col_children_metadata(dtype,
+                                column_metadata& col_meta):
+    if is_struct_dtype(dtype):
+        col_meta.children_meta.reserve(len(dtype.fields))
+        for i, name in enumerate(dtype.fields):
+            value = dtype.fields[name]
+            col_meta.children_meta.push_back(column_metadata(name.encode()))
+            _set_col_children_metadata(
+                value, col_meta.children_meta[i]
+            )
+    elif is_list_dtype(dtype):
+        pass
     else:
-        raise ValueError("Malformed metadata has been encountered")
+        return
```

Comment:
Passing here implicitly returns `None`, which is the same as a bare `return`. But is any of this code needed? Seems like all 4 lines can be deleted (I can't put a GitHub suggestion here because it overlaps with deleted lines).